scico/linop/_diag.py

```diff
@@ -156,10 +156,137 @@ def norm(self, ord=None):  # pylint: disable=W0622
+        return self._diagonal * x
+
+    @property
+    def diagonal(self) -> Union[Array, BlockArray]:
+        return self._diagonal * snp.ones(self.input_shape, dtype=self.input_dtype)
+
+    def conj(self) -> ScaledIdentity:
+        """Complex conjugate of this :class:`ScaledIdentity`."""
+        return ScaledIdentity(
+            scalar=self._diagonal.conj(), input_shape=self.input_shape, input_dtype=self.input_dtype
+        )
+
+    @property
+    def gram_op(self) -> ScaledIdentity:
+        """Gram operator of this :class:`ScaledIdentity`."""
+        return ScaledIdentity(
+            scalar=self._diagonal * self._diagonal.conj(),
+            input_shape=self.input_shape,
+            input_dtype=self.input_dtype,
+        )
+
+    @partial(_wrap_add_sub, op=operator.add)
+    def __add__(self, other):
+        if self.input_shape == other.input_shape:
+            return ScaledIdentity(
+                scalar=self._diagonal + other._diagonal,
+                input_shape=self.input_shape,
+                input_dtype=self.input_dtype,
+            )
+        raise ValueError(f"Incompatible shapes: {self.shape} != {other.shape}.")
+
+    @partial(_wrap_add_sub, op=operator.sub)
+    def __sub__(self, other):
+        if self.input_shape == other.input_shape:
+            return ScaledIdentity(
+                scalar=self._diagonal - other._diagonal,
+                input_shape=self.input_shape,
+                input_dtype=self.input_dtype,
+            )
+        raise ValueError(f"Incompatible shapes: {self.shape} != {other.shape}.")
+
+    @_wrap_mul_div_scalar
+    def __mul__(self, scalar):
+        return ScaledIdentity(
+            scalar=self._diagonal * scalar,
+            input_shape=self.input_shape,
+            input_dtype=self.input_dtype,
+        )
+
+    @_wrap_mul_div_scalar
+    def __rmul__(self, scalar):
+        return ScaledIdentity(
+            scalar=self._diagonal * scalar,
+            input_shape=self.input_shape,
+            input_dtype=self.input_dtype,
+        )
+
+    @_wrap_mul_div_scalar
+    def __truediv__(self, scalar):
+        return ScaledIdentity(
+            scalar=self._diagonal / scalar,
+            input_shape=self.input_shape,
+            input_dtype=self.input_dtype,
+        )
+
+    def __matmul__(self, other):
+        # self @ other
+        if isinstance(other, Diagonal):
+            if self.shape != other.shape:
+                raise ValueError(f"Shapes {self.shape} and {other.shape} do not match.")
+            if isinstance(other, ScaledIdentity):
+                return ScaledIdentity(
+                    scalar=self._diagonal * other._diagonal,
+                    input_shape=self.input_shape,
+                    input_dtype=self.input_dtype,
+                )
+            else:
+                return Diagonal(diagonal=self._diagonal * other.diagonal)
+        else:
+            return self(other)
```

Comment:
Not a problem with this PR, but the previous 50+ lines of code to express `scalar * x` is a code smell. Let's make an issue (or dev note) about looking into these wrappers.

Reply:
Agreed. Please add a note to #502 if what's already there doesn't fully reflect what you have in mind.